### PR TITLE
API: Added coverage for location update

### LINKED ISSUE #2427: Automate location update in api
please automate

PUT /api/locations/:id
Update a location
Params
Param name 	Description
location
required 	
Value: Hash
location[name]
optional 	
Value: String
location[description]
optional , nil allowed 	
Value: String
location[user_ids]
optional , nil allowed 	

User IDs

Value: Must be an array of any type
location[smart_proxy_ids]
optional , nil allowed 	

Smart proxy IDs

Value: Must be an array of any type
location[compute_resource_ids]
optional , nil allowed 	

Compute resource IDs

Value: Must be an array of any type
location[media_ids]
optional , nil allowed 	

Media IDs

Value: Must be an array of any type
location[config_template_ids]
optional , nil allowed 	

Provisioning template IDs

Value: Must be an array of any type
location[domain_ids]
optional , nil allowed 	

Domain IDs

Value: Must be an array of any type
location[realm_ids]
optional , nil allowed 	

Realm IDs

Value: Must be an array of any type
location[hostgroup_ids]
optional , nil allowed 	

Host group IDs

Value: Must be an array of any type
location[environment_ids]
optional , nil allowed 	

Environment IDs

Value: Must be an array of any type
location[subnet_ids]
optional , nil allowed 	

Subnet IDs

Value: Must be an array of any type

### PR DESCRIPTION
**Pull request by abalakh**
Added coverage for ```PUT /api/locations/:id```.
Closes #2427
Tests results:
```
nosetests tests/foreman/api/test_location.py -m test_update                                          
................................
----------------------------------------------------------------------
Ran 32 tests in 201.256s

OK
```
```[DO NOT MERGE]``` tag until SatelliteQE/nailgun#130 gets merged and released.

### CODE REVIEW

File: tests/foreman/api/test_location.py [hunk 1]
```diff
@@ -199,3 +200,276 @@ def test_create_location_with_multiple_organization(self):
...
+            config_template=[entities.ConfigTemplate().create()],
+        ).create()
+        template = entities.ConfigTemplate().create()
+        location.config_template = [template]
+        ct_list = [
+            ct
+            for ct
+            in location.update(['config_template']).config_template
+            if ct.id == template.id
+        ]
```
Comment (Ichimonji10) [inline tests/foreman/api/test_location.py]: :+1: For using this instead of filter and lambda.

File: tests/foreman/api/test_location.py [hunk 2]
```diff
@@ -199,3 +200,276 @@ def test_create_location_with_multiple_organization(self):
...
+        ).create()
+        orgs_amount = randint(3, 5)
+        orgs = [
+            entities.Organization().create() for _ in range(orgs_amount)
+        ]
+        location.organization = orgs
+        location = location.update(['organization'])
+        self.assertEqual(len(location.organization), orgs_amount)
+        for organization in location.organization:
+            self.assertIn(organization.id, [org.id for org in orgs])
```
Comment (Ichimonji10) [inline tests/foreman/api/test_location.py]: unittest provides great error messages when an assertion is made against a set. For example, if I wrote this:

```python
self.assertEqual(set([1, 2, 3]), set([2, 3, 4]))
```

I would get an error message something like this:

    Items in the first set but not the second: [1]
    Items in the second set but not the first: [4]

With that in mind, here's an alternate form for this test:

```python
        location = entities.Location(
            organization=[entities.Organization().create()],
        ).create()
        orgs = [
            entities.Organization().create() for _ in range(randint(3, 5))
        ]
        location.organization = orgs
        location = location.update(['organization'])
        self.assertEqual(
            set([org.id for org in orgs]),
            set([org.id for org in location.organization]),
        )
```

**Comment by Ichimonji10** [comment]
Nice job. ACK pending comment.

**Comment by abalakh** [comment]
Rebased and updated according to comment

File: tests/foreman/api/test_location.py (hunk 2 above)
Comment (abalakh) [inline tests/foreman/api/test_location.py]: :+1:

File: tests/foreman/api/test_location.py [hunk 3]
```diff
@@ -213,3 +213,274 @@ def test_delete_location(self):
...
+        """
+        location = entities.Location(
+            organization=[entities.Organization().create()],
+        ).create()
+        orgs = [entities.Organization().create() for _ in range(randint(3, 5))]
+        location.organization = orgs
+        location = location.update(['organization'])
+        self.assertEqual(
+            set([org.id for org in orgs]),
+            set([org.id for org in location.organization]),
```
Comment (jyejare) [inline tests/foreman/api/test_location.py]: Nice use of set for filtering the elements for comparison.
@abalakh ++

**Comment by jyejare** [comment]
PR looks nice. Thanks to Nailgun for making things simple.
ACK.

**Comment by oshtaier** [comment]
Third ACK

File: tests/foreman/api/test_location.py [hunk 4]
```diff
@@ -213,3 +213,274 @@ def test_delete_location(self):
         location.delete()
         with self.assertRaises(HTTPError):
             location.read()
+
+    @data(
+        gen_string('alphanumeric', randint(1, 255)),
```
Comment (oshtaier) [inline tests/foreman/api/test_location.py]: Change value according to #2525

**Comment by Ichimonji10** [comment]
Bonus karma for rebasing this. The parent commit is 7 days old.

**Comment by abalakh** [comment]
Sure, done :)

**Pull request merged**